### PR TITLE
feat: dpop support

### PR DESCRIPTION
**Pull request by auer-martin**
blocked by for now https://github.com/Sphereon-Opensource/OID4VC/pull/131

### CODE REVIEW

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 1]
```diff
@@ -263,6 +273,44 @@ export class OpenId4VciHolderService {
...
+    }
+  ) {
+    const dPoPSigningAlgValuesSupported =
+      metadata.authorizationServerMetadata?.dpop_signing_alg_values_supported ??
+      metadata.credentialIssuerMetadata.dpop_signing_alg_values_supported
+
+    if (!dPoPSigningAlgValuesSupported) return undefined
+
+    const alg = dPoPSigningAlgValuesSupported
+      .flatMap((alg) => alg)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:289]: Why is this needed? It's already string array right?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 2]
```diff
@@ -271,6 +319,11 @@ export class OpenId4VciHolderService {
     let accessTokenResponse: OpenIDResponse<AccessTokenResponse>
 
     const accessTokenClient = new AccessTokenClient()
+
+    const createDPoPOptions = await this.getCreateDPoPOptions(agentContext, metadata)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:323]: In credo we usually remove the capitalization if it's being used aa a word.

So createDpopOptions. Then you always know when a word starts and stops.

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 3]
```diff
@@ -271,6 +319,11 @@ export class OpenId4VciHolderService {
...
+    const dPoPJwk = createDPoPOptions ? getJwkFromJson(createDPoPOptions.jwtIssuer.jwk) : undefined
+
+    resolvedCredentialOffer.metadata.credentialIssuerMetadata
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:326]:
```suggestion
```

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 4]
```diff
@@ -297,7 +352,7 @@ export class OpenId4VciHolderService {
 
     this.logger.debug('Requested OpenId4VCI Access Token.')
 
-    return accessTokenResponse.successBody
+    return { ...accessTokenResponse.successBody, dPoPJwk }
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:355]: This is infortunate, but i don't think we should mix it. Ideally we return a new object with keys: {response: {body,headers,code}, dpop: {jwk}}


Then It's more extendable. 

Maybe deprecate and add a new one?  Or we keep it like this for now and upgrade when moving to 0.6


But still would like to have dpop.jwk so we can extend dpop values as well.

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 5]
```diff
@@ -308,6 +363,7 @@ export class OpenId4VciHolderService {
       resolvedAuthorizationRequestWithCode?: OpenId4VciResolvedAuthorizationRequestWithCode
       accessToken?: string
       cNonce?: string
+      dPoPJwk?: Jwk
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:366]: dpopJwk

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 6]
```diff
@@ -347,7 +405,7 @@ export class OpenId4VciHolderService {
     } as OpenId4VciTokenRequestOptions
 
     const tokenResponse = options.accessToken
-      ? { access_token: options.accessToken, c_nonce: options.cNonce }
+      ? { access_token: options.accessToken, c_nonce: options.cNonce, dPoPJwk: options.dPoPJwk }
       : await this.requestAccessToken(agentContext, tokenRequestOptions)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:409]: For the service we can already change it (only api we need to worry about breaking changes)

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 7]
```diff
@@ -412,10 +470,24 @@ export class OpenId4VciHolderService {
         .withToken(tokenResponse.access_token)
 
       const credentialRequestClient = credentialRequestBuilder.build()
+
+      let createDPoPOptions: CreateDPoPClientOptions | undefined
+      if (options.dPoPJwk) {
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:475]: Isn't the dpop jwk also included in the access token cnf? In that case, can't we dynamically detect it and then use dpop?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderServiceOptions.ts [hunk 8]
```diff
@@ -112,6 +112,7 @@ export interface OpenId4VciCredentialRequestOptions extends Omit<OpenId4VciAccep
   resolvedCredentialOffer: OpenId4VciResolvedCredentialOffer
   accessToken: string
   cNonce?: string
+  dPoPJwk?: Jwk
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderServiceOptions.ts:115]: dpopJwk ( you get it)

File: packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerApi.ts [hunk 9]
```diff
@@ -76,6 +76,7 @@ export class OpenId4VcIssuerApi {
       issuer.credentialConfigurationsSupported = undefined
     }
     issuer.display = options.display
+    issuer.dPoPSigningAlgValuesSupported = options.dPoPSigningAlgValuesSupported
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerApi.ts:79]: I feel like we'll have more issuer metadata properties over time that we want to add. Should we have a generic way for it?

File: packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerService.ts [hunk 10]
```diff
@@ -303,6 +303,7 @@ export class OpenId4VcIssuerService {
     const openId4VcIssuerBase = {
       issuerId: options.issuerId ?? utils.uuid(),
       display: options.display,
+      dPoPSigningAlgValuesSupported: options.dPoPSigningAlgValuesSupported,
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerService.ts:306]:
```suggestion
      dpopSigningAlgValuesSupported: options.dPoPSigningAlgValuesSupported,
```

File: packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts [hunk 11]
```diff
@@ -120,6 +122,30 @@ export function handleTokenRequest(config: OpenId4VciAccessTokenEndpointConfig)
     const issuerMetadata = openId4VcIssuerService.getIssuerMetadata(agentContext, issuer)
     const accessTokenSigningKey = Key.fromFingerprint(issuer.accessTokenPublicKeyFingerprint)
 
+    let dPoPJwk: JWK | undefined
+    if (request.headers.dpop) {
+      try {
+        const fullUrl = request.protocol + '://' + request.get('host') + request.originalUrl
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts:128]: Need to use the base url from module configuration. This will break

File: packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts [hunk 12]
```diff
@@ -120,6 +122,30 @@ export function handleTokenRequest(config: OpenId4VciAccessTokenEndpointConfig)
     const issuerMetadata = openId4VcIssuerService.getIssuerMetadata(agentContext, issuer)
     const accessTokenSigningKey = Key.fromFingerprint(issuer.accessTokenPublicKeyFingerprint)
 
+    let dPoPJwk: JWK | undefined
+    if (request.headers.dpop) {
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts:126]: Can we require dpop?

File: packages/openid4vc/src/openid4vc-issuer/router/verifyResourceRequest.ts [hunk 13]
```diff
@@ -27,6 +36,15 @@ export async function verifyAccessToken(
     },
   })
 
+  const fullUrl = request.protocol + '://' + request.get('host') + request.originalUrl
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/router/verifyResourceRequest.ts:39]: Same here . Need to use baseUrl from configuration

File: packages/openid4vc/src/openid4vc-issuer/router/verifyResourceRequest.ts [hunk 14]
```diff
@@ -27,6 +36,15 @@ export async function verifyAccessToken(
...
+  await verifyResourceDPoP(
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/router/verifyResourceRequest.ts:40]: I know it looks a bit weird, but consistency is king imo

```suggestion
  await verifyResourceDpop(
```

File: packages/openid4vc/src/shared/utils.ts [hunk 15]
```diff
@@ -124,13 +134,20 @@ export async function openIdTokenIssuerToJwtIssuer(
...
+    }
+
     if (!issuer.startsWith('https://')) {
       throw new CredoError('The X509 certificate issuer must be a HTTPS URI.')
     }
 
     if (leafCertificate.sanUriNames?.includes(issuer)) {
       return {
         ...openId4VcTokenIssuer,
+        alg: alg as unknown as SigningAlgo,
```
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:150]: I think the lib should just support string. As you can use any algorithm and this is a bit annoying to cast everytime

File: packages/openid4vc/tests/openid4vc.e2e.test.ts [hunk 16]
```diff
@@ -218,6 +223,7 @@ describe('OpenId4Vc', () => {
       },
     })
     const openIdIssuerTenant2 = await issuerTenant2.modules.openId4VcIssuer.createIssuer({
+      dPoPSigningAlgValuesSupported: ['EdDSA'],
```
Comment (TimoGlastra) [inline packages/openid4vc/tests/openid4vc.e2e.test.ts:226]: Thinking in general whether this is something you want to configure per issuer? It will make uodates in multi-tenancy quite cumbersome. But i also see the case for having different values per issuer

**Review by TimoGlastra — COMMENTED** [review]
Nice very good stuff martin!!'

Only thinkg i'm not sure about is the retuen value of requestAccesToken. Bit unfortunate we already released the previous pr as now it will be a breaking change...

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 7 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:475]: no

File: packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts (hunk 12 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts:126]: Well, we could implement it. But in theory, the client can opt into it for more security.
Also, supporting dpop without having clients use it is perfectly viable.
If we implement it we should probably create a config field allowing users to opt into that strict behaviour

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 4 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:355]: I would leave it as is for now. Maybe change with 0.6 then.

> But still would like to have dpop.jwk so we can extend dpop values as well. 
On the client side you only receive the dpop not really things that can be extended IMO

Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:352]: Until we maybe want to return the dpop that was sent, or the htu, or anything.. I think we've had a few times now where not adding an extra layer of nesting resulted in using having to make breaking changes. Rather take that into account now

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 7 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:472]: Interesting. For sd-jwt they include the jwk directly but here a fingerpring...

File: packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts (hunk 12 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-issuer/router/accessTokenEndpoint.ts:136]: For the funke pid issuer it is a requirement. There's going to be cases where the issuer wants to require it (fine to keep it out for now, but something we can add later. Maybe on the issuer record, or per credential offer)

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 17]
```diff
@@ -266,14 +275,60 @@ export class OpenId4VciHolderService {
...
 
     const accessTokenClient = new AccessTokenClient()
+
+    const createDPoPOpts = await this.getCreateDPoPOptions(agentContext, metadata)
+
+    let dpopJwk: Jwk | undefined
+    if (createDPoPOpts) {
+      if (!createDPoPOpts.jwtIssuer.jwk.kty) {
+        throw new CredoError('Missing required key type (kty) in the jwk.')
+      }
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:329]: why is this needed? In theory the jwk should have this property, so I'm not sure why we check for this specific property

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 18]
```diff
@@ -266,14 +275,60 @@ export class OpenId4VciHolderService {
...
+      dpopJwk = getJwkFromJson(createDPoPOpts.jwtIssuer.jwk as JwkJson)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:330]: Why first transform from JWK class to json and then back to jwk, shouldn't the createDpopOptions just return the instance?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts [hunk 19]
```diff
@@ -146,11 +146,12 @@ export class OpenId4VcHolderApi {
...
-    const { access_token: accessToken, c_nonce: cNonce } = await this.openId4VciHolderService.requestAccessToken(
-      this.agentContext,
-      options
-    )
-    return { accessToken, cNonce }
+    const {
+      access_token: accessToken,
+      c_nonce: cNonce,
+      dpop,
+    } = await this.openId4VciHolderService.requestAccessToken(this.agentContext, options)
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts:153]: We can change the service interface just fine, so let's change the service return value to:
```
{
  accessToken,
  cNonce,
  dpop,
  response // should be cloned response (using .clone)
}
```

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 20]
```diff
@@ -311,6 +371,7 @@ export class OpenId4VciHolderService {
       resolvedAuthorizationRequestWithCode?: OpenId4VciResolvedAuthorizationRequestWithCode
       accessToken?: string
       cNonce?: string
+      dpop?: { dpopJwk: Jwk; dpopNonce?: string }
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:374]: just thinking: is it needed to prefix dpop if we already are in `dpop` object?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 21]
```diff
@@ -351,7 +414,11 @@ export class OpenId4VciHolderService {
     } as OpenId4VciTokenRequestOptions
 
     const tokenResponse = options.accessToken
-      ? { access_token: options.accessToken, c_nonce: options.cNonce }
+      ? {
+          access_token: options.accessToken,
+          c_nonce: options.cNonce,
+          ...(options.dpop && { dpop: { dpopJwk: options.dpop.dpopJwk, dpopNonce: options.dpop?.dpopNonce } }),
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:420]: this won't work?

```suggestion
          ...(options.dpop && { dpop: options.dpop  }),
```

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts [hunk 22]
```diff
@@ -425,10 +492,24 @@ export class OpenId4VciHolderService {
...
+      if (tokenResponse.dpop) {
+        const jwk = tokenResponse.dpop.dpopJwk
+        const alg = jwk.supportedSignatureAlgorithms[0]
+
+        createDpopOpts = {
+          jwtIssuer: { alg: alg as unknown as SigningAlgo, jwk: jwk.toJson() },
+          jwtPayloadProps: { accessToken: tokenResponse.access_token, nonce: tokenResponse.dpop?.dpopNonce },
+          createJwtCallback: getCreateJwtCallback(agentContext),
+        }
+      }
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:506]: Why not reuse `this.getCreateDPoPOptions` here? Also this could give an error when multiple alg values are supported for a jwk type, but the server only supports a specific one. So reusing the dpop alg values supported logic from existing method makes sense?

Also, accessing index [0] is unsafe again

Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:506]: Also should we check if dpop is required? Can we see that using the access token? Or how do we know it is required?

File: packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerApi.ts [hunk 23]
```diff
@@ -63,21 +63,22 @@ export class OpenId4VcIssuerApi {
...
+      issuer.credentialsSupported = credentialsSupportedV13ToV11(credentialConfigurationsSupported)
     } else {
-      issuer.credentialsSupported = options.credentialsSupported
+      issuer.credentialsSupported = credentialsSupported
       issuer.credentialConfigurationsSupported = undefined
     }
-    issuer.display = options.display
 
-    return this.openId4VcIssuerService.updateIssuer(this.agentContext, issuer)
+    return this.openId4VcIssuerService.updateIssuer(this.agentContext, Object.assign(issuer, issuerOptions))
```
Comment (TimoGlastra) [inline packages/openid4vc/src/openid4vc-issuer/OpenId4VcIssuerApi.ts:81]: I'd rather manually assign the props that can be updated instead of doing object.assign. Especially since if you're not using TypeScript it will just assign any property to the record that you pass here. You could provide e.g. `type: 'Random'` and it will overwrite the type of the record.

File: packages/openid4vc/src/shared/utils.ts [hunk 24]
```diff
@@ -66,23 +67,29 @@ export function getVerifyJwtCallback(agentContext: AgentContext): VerifyJwtCallb
...
         payload: JwtPayload.fromJson(jwt.payload),
         key,
       })
 
       return jws
     } else if (jwtIssuer.method === 'jwk') {
-      const key = getJwkFromJson(jwtIssuer.jwk).key
+      if (!jwtIssuer.jwk.kty) {
+        throw new CredoError('Missing required key type (kty) in the jwk.')
+      }
```
Comment (TimoGlastra) [inline packages/openid4vc/src/shared/utils.ts:88]: again, why this check?

**Review by TimoGlastra — APPROVED** [review]

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 18 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:330]: Hmmm. I don't think so because we create the options which are passed to sphereons library and there we don't have the jwk class

File: packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts (hunk 19 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts:153]: ? what should the response contain? the whole response?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 22 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts:506]: if the tokenresponse contains dpop it is required

File: packages/openid4vc/src/shared/utils.ts (hunk 24 above)
Comment (auer-martin) [reply packages/openid4vc/src/shared/utils.ts:88]: can be undefined in sphereon

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 22 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts]: But token response is passed by the user

File: packages/openid4vc/src/shared/utils.ts (hunk 24 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/shared/utils.ts:88]: But in this case we have a jwk class instance right? So it won't happen?

File: packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts (hunk 19 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts:153]: Yes it would be a start at solving https://github.com/openwallet-foundation/credo-ts/issues/1965, which was requested. Returning the response means better extensibility as you can extract the headers, etc..

Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts:153]: In the PR where I implemented the tokenRequestfunctionality, I had that behaviour, and you made me change it.

File: packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts (hunk 22 above)
Comment (auer-martin) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VciHolderService.ts]: Hmm, IMO, the user should not modify the object unless he knows what he is doing.
Also in theory it is valid to remove it since an authorization server can protect multiple credential issuers,
for some of them use of dpop may be optional even if the authorization server requires dpop.

File: packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts (hunk 19 above)
Comment (TimoGlastra) [reply packages/openid4vc/src/openid4vc-holder/OpenId4VcHolderApi.ts:153]: Okay then i misunderstood the previous implementation. What i don't want is requiring the whole access token response as input to the request credentials method.

I want to have separation between the required fields for credo to work, and additional params you may need for extension. And as I said in the previous PR, i'm not a fan of mixing the successBody with extra params we extracted from e.g. the header.

**Comment by TimoGlastra** [comment]
Made a small mistake in my PR on jwk thumbprint, so we need to wait for https://github.com/Sphereon-Opensource/OID4VC/pull/140
